contracts/DCAStrategies/DCAStrategies/DCAStrategiesPositionsHandler.sol

```diff
@@ -189,7 +197,63 @@ abstract contract DCAStrategiesPositionsHandler is IDCAStrategiesPositionsHandle
     Position memory _position = _userPositions[_positionId];
     IDCAStrategies.ShareOfToken[] memory _newTokenShares = _getTokenShares(_position.strategyId, _newVersion);
 
-    _syncFirstBlock(_position, _totalAmount, _newTokenShares, _newAmountSwaps, _recipientSwapped);
+    (Data memory _data, Task[] memory _tasks) = _syncBlock(_position, _totalAmount, _newTokenShares, _newAmountSwaps, _recipientSwapped);
+
+    // If get to this place, then we just need to terminate existing positions
+    while (_data.currentPositionsIndex < _position.positions.length) {
+      (uint256 _unswapped, ) = _position.hub.terminate(_position.positions[_data.currentPositionsIndex], address(this), _recipientSwapped);
+      _data.totalRemaining += _unswapped;
+    }
+
+    while (_data.newPositionsIndex < _newTokenShares.length) {
+      IDCAStrategies.ShareOfToken memory _newTokenShare = _newTokenShares[_data.newPositionsIndex];
+      uint256 _correspondingToPosition = _calculateOptimalAmount(
+        _data.newPositionsIndex == _newTokenShares.length - 1,
+        _totalAmount,
+        _newTokenShare.share,
+        _getTotalShares(),
+        _data.amountSpent
+      );
+      _tasks[_data.newPositionsIndex] = Task({action: Action.DEPOSIT, amount: _correspondingToPosition, positionId: _positionId});
+      _data.amountSpent += _correspondingToPosition;
+    }
+
+    // extract (increase) or send (reduce)
+    if (_totalAmount > _data.totalRemaining) {
+      IERC20(_data.positionMetadata.from).safeTransferFrom(msg.sender, address(this), _totalAmount - _data.totalRemaining);
+    } else if (_totalAmount < _data.totalRemaining) {
+      IERC20(_data.positionMetadata.from).safeTransfer(_recipientUnswapped, _data.totalRemaining - _totalAmount);
+    }
+
+    // perform deposit and increase
+    for (uint256 i = 0; i < _tasks.length; ) {
+      Task memory _task = _tasks[i];
+
+      if (_task.action == Action.INCREASE) {
+        _position.hub.increasePosition(_task.positionId, _task.amount, _newAmountSwaps);
+      } else if (_task.action == Action.DEPOSIT) {
+        uint256 _newPositionId = _position.hub.deposit(
+          address(_data.positionMetadata.from),
+          _newTokenShares[i].token,
+          _task.amount,
+          _newAmountSwaps,
+          _data.positionMetadata.swapInterval,
+          address(this),
+          new IDCAPermissionManager.PermissionSet[](0)
+        );
+        if (i < _position.positions.length) {
+          _userPositions[_task.positionId].positions[i] = _newPositionId;
```

Comment:
Let's add a TODO so we can see if there are other ways we can modify the function and use the `_positionId` from the parameters, without a `stack too deep error`